### PR TITLE
Boson voucher - after audit remediations

### LINKED ISSUE #500: [BVR-01S] Inexistent Conformity to the Checks-Effects-Interactions Pattern
## BVR-01S: Inexistent Conformity to the Checks-Effects-Interactions Pattern

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Language Specific](/appendix/finding-types#language-specific) |  | BosonVoucher.sol:L645-L649, [L650](https://github.com/bosonprotocol/boson-protocol-contracts/blob/38c7ba384e55486aee4ea76bb80b4f29d1a86175/contracts/protocol/clients/voucher/BosonVoucher.sol#L650) |

### Description:

The `_beforeTransferHook` function will first invoke `commitToPreMintedOffer` prior to issuing a `delete` operation on the `_premintStatus` struct, permitting a re-entrancy attack to potentially occur within a consequent call of `commitToPreMintedOffer`.

### Impact:

The Checks-Effects-Interactions (CEI) pattern should always be applied in all function implementations to ensure no latent re-entrancy attacks are present that can manifest in a future update of the protocol that i.e. introduces unsafe calls to `commitToPreMintedOffer`. This scenario is not unlikely as one can assume that it is protected by the `nonReentrant` modifier whilst contracts interacting with it would be susceptible to re-entrancies.

### Example:

```solidity
function _beforeTokenTransfer(
    address _from,
    address _to,
    uint256 _tokenId
) internal override {
    // Update the buyer associated with the voucher in the protocol
    // Only when transferring, not minting or burning
    if (_from == owner()) {
        if (_premintStatus.committable) {
            // Set the preminted token as committed
            _committed[_tokenId] = true;

            // If this is a transfer of preminted token, treat it differently
            address protocolDiamond = IClientExternalAddresses(BeaconClientLib._beacon()).getProtocolAddress();
            IBosonExchangeHandler(protocolDiamond).commitToPreMintedOffer(
                payable(_to),
                _premintStatus.offerId,
                _tokenId
            );
            delete _premintStatus;
        } else {
            // Already committed, treat as a normal transfer
            onVoucherTransferred(_tokenId, payable(_to));
        }
    } else if (_from != address(0) && _to != address(0) && _from != _to) {
        onVoucherTransferred(_tokenId, payable(_to));
    }
}
```

### Recommendation:

Although in the current implementation `commitToPreMintedOffer` is `nonReentrant` and the function solely invokes read-only functions of untrusted contracts, no security vulnerability arises from the mis-applied pattern. In any case, we advise the `delete` instruction to be relocated prior to the external calls as the `_premintStatus` is not utilized in between.

### LINKED ISSUE #501: [BVR-01M] Inexistent Sanitization of Range Validity
## BVR-01M: Inexistent Sanitization of Range Validity

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Mathematical Operations](/appendix/finding-types#mathematical-operations) |  | BosonVoucher.sol:L168, [L169](https://github.com/bosonprotocol/boson-protocol-contracts/blob/38c7ba384e55486aee4ea76bb80b4f29d1a86175/contracts/protocol/clients/voucher/BosonVoucher.sol#L169) |

### Description:

The `reserveRange` function does not explicitly validate that the range defined can be adequately computed (i.e. does not overflow via addition in `issueVoucher`).

### Impact:

The issue does not pose an active threat, however, this validation should be enforced locally to ensure the security guarantees of the `BosonVoucher` contract are upheld self-sufficiently.

### Example:

```solidity
/**
 * @notice Initializes the voucher.
 * This function is callable only once.
 */
function initializeVoucher(
    uint256 _sellerId,
    address _newOwner,
    VoucherInitValues calldata voucherInitValues
) public initializer {
    string memory sellerId = Strings.toString(_sellerId);
    string memory voucherName = string(abi.encodePacked(VOUCHER_NAME, " ", sellerId));
    string memory voucherSymbol = string(abi.encodePacked(VOUCHER_SYMBOL, "_", sellerId));

    __ERC721_init_unchained(voucherName, voucherSymbol);

    // we don't call init on ownable, but rather just set the ownership to correct owner
    _transferOwnership(_newOwner);

    _setContractURI(voucherInitValues.contractURI);

    _setRoyaltyPercentage(voucherInitValues.royaltyPercentage);

    emit VoucherInitialized(_sellerId, _royaltyPercentage, _contractURI);
}

/**
 * @notice Issues a voucher to a buyer.
 *
 * Minted voucher supply is sent to the buyer.
 * Caller must have PROTOCOL role.
 *
 * Reverts if:
 * - Exchange id falls within a reserved range
 *
 * @param _exchangeId - the id of the exchange (corresponds to the ERC-721 token id)
 * @param _buyer - the buyer address
 */
function issueVoucher(uint256 _exchangeId, address _buyer) external override onlyRole(PROTOCOL) {
    // Get the exchange
    (, Exchange memory exchange) = getBosonExchange(_exchangeId);

    // See if the offer id is associated with a range
    Range storage range = _rangeByOfferId[exchange.offerId];

    // Revert if exchange id falls within a reserved range
    uint256 rangeStart = range.start;
    require(
        (_exchangeId < rangeStart) || (_exchangeId >= rangeStart + range.length),
        EXCHANGE_ID_IN_RESERVED_RANGE
    );

    // Mint the voucher, sending it to the buyer address
    _mint(_buyer, _exchangeId);
}

/**
 * @notice Burns a voucher.
 *
 * Caller must have PROTOCOL role.
 *
 * @param _exchangeId - the id of the exchange (corresponds to the ERC-721 token id)
 */
function burnVoucher(uint256 _exchangeId) external override onlyRole(PROTOCOL) {
    _burn(_exchangeId);
}

/**
 * @notice Reserves a range of vouchers to be associated with an offer
 *
 * Must happen prior to calling preMint
 * Caller must have PROTOCOL role.
 *
 * Reverts if:
 * - Start id is not greater than zero
 * - Offer id is already associated with a range
 *
 * @param _offerId - the id of the offer
 * @param _start - the first id of the token range
 * @param _length - the length of the range
 */
function reserveRange(
    uint256 _offerId,
    uint256 _start,
    uint256 _length
) external onlyRole(PROTOCOL) {
    // Make sure range start is valid
    require(_start > 0, INVALID_RANGE_START);

    // Get storage slot for the range
    Range storage range = _rangeByOfferId[_offerId];

    // Revert if the offer id is already associated with a range
    require(range.length == 0, OFFER_RANGE_ALREADY_RESERVED);

    // Store the reserved range
    range.offerId = _offerId;
    range.start = _start;
    range.length = _length;
    _rangeOfferIds.push(_offerId);

    emit RangeReserved(_offerId, range);
}
```

### Recommendation:

While this is indirectly protected against via the maximum length limitation wherever `reserveRange` is invoked, it should also be validated locally as it is currently possible to create an invalid range entry which will be unusable due to an overflow failure.

### LINKED ISSUE #502: [BVR-02M] Inexistent Validation of Range Length
## BVR-02M: Inexistent Validation of Range Length

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Input Sanitization](/appendix/finding-types#input-sanitization) |  | BosonVoucher.sol:L119, [L155](https://github.com/bosonprotocol/boson-protocol-contracts/blob/38c7ba384e55486aee4ea76bb80b4f29d1a86175/contracts/protocol/clients/voucher/BosonVoucher.sol#L155) |

### Description:

The implementation of `issueVoucher` assumes that at least one member of the range will be reserved based on the first comparator of the newly introduced `require` check, however, the `reserveRange` function does not enforce a minimum `_length`. While the function is protected from an incorrect `_length` variable at the invocation level within `OfferBase`, it is also protected against an invalid `_start` even though it is validated locally.

### Impact:

While currently not an active issue due to `reserveRange` being invoked by properly defined functions, this issue can arise in a future integration by another module of the Boson system invoking `reserveRange`.

### Example:

```solidity
/**
 * @notice Initializes the voucher.
 * This function is callable only once.
 */
function initializeVoucher(
    uint256 _sellerId,
    address _newOwner,
    VoucherInitValues calldata voucherInitValues
) public initializer {
    string memory sellerId = Strings.toString(_sellerId);
    string memory voucherName = string(abi.encodePacked(VOUCHER_NAME, " ", sellerId));
    string memory voucherSymbol = string(abi.encodePacked(VOUCHER_SYMBOL, "_", sellerId));

    __ERC721_init_unchained(voucherName, voucherSymbol);

    // we don't call init on ownable, but rather just set the ownership to correct owner
    _transferOwnership(_newOwner);

    _setContractURI(voucherInitValues.contractURI);

    _setRoyaltyPercentage(voucherInitValues.royaltyPercentage);

    emit VoucherInitialized(_sellerId, _royaltyPercentage, _contractURI);
}

/**
 * @notice Issues a voucher to a buyer.
 *
 * Minted voucher supply is sent to the buyer.
 * Caller must have PROTOCOL role.
 *
 * Reverts if:
 * - Exchange id falls within a reserved range
 *
 * @param _exchangeId - the id of the exchange (corresponds to the ERC-721 token id)
 * @param _buyer - the buyer address
 */
function issueVoucher(uint256 _exchangeId, address _buyer) external override onlyRole(PROTOCOL) {
    // Get the exchange
    (, Exchange memory exchange) = getBosonExchange(_exchangeId);

    // See if the offer id is associated with a range
    Range storage range = _rangeByOfferId[exchange.offerId];

    // Revert if exchange id falls within a reserved range
    uint256 rangeStart = range.start;
    require(
        (_exchangeId < rangeStart) || (_exchangeId >= rangeStart + range.length),
        EXCHANGE_ID_IN_RESERVED_RANGE
    );

    // Mint the voucher, sending it to the buyer address
    _mint(_buyer, _exchangeId);
}

/**
 * @notice Burns a voucher.
 *
 * Caller must have PROTOCOL role.
 *
 * @param _exchangeId - the id of the exchange (corresponds to the ERC-721 token id)
 */
function burnVoucher(uint256 _exchangeId) external override onlyRole(PROTOCOL) {
    _burn(_exchangeId);
}

/**
 * @notice Reserves a range of vouchers to be associated with an offer
 *
 * Must happen prior to calling preMint
 * Caller must have PROTOCOL role.
 *
 * Reverts if:
 * - Start id is not greater than zero
 * - Offer id is already associated with a range
 *
 * @param _offerId - the id of the offer
 * @param _start - the first id of the token range
 * @param _length - the length of the range
 */
function reserveRange(
    uint256 _offerId,
    uint256 _start,
    uint256 _length
) external onlyRole(PROTOCOL) {
    // Make sure range start is valid
    require(_start > 0, INVALID_RANGE_START);

    // Get storage slot for the range
    Range storage range = _rangeByOfferId[_offerId];

    // Revert if the offer id is already associated with a range
    require(range.length == 0, OFFER_RANGE_ALREADY_RESERVED);

    // Store the reserved range
    range.offerId = _offerId;
    range.start = _start;
    range.length = _length;
    _rangeOfferIds.push(_offerId);

    emit RangeReserved(_offerId, range);
}
```

### Recommendation:

We advise variable validation to be entirely relocated within the `BosonVoucher` to ensure self-sufficiency and a `require` check to be introduced ensuring a non-zero `_length` argument.

### LINKED ISSUE #503: [BVR-03M] Weak Enforcement of `_rangeOfferIds` Ascending Order
## BVR-03M: Weak Enforcement of `_rangeOfferIds` Ascending Order

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) |  | BosonVoucher.sol:L152-L173 |

### Description:

The `_rangeOfferIds` array entries should point to ranges that are in ascending order for the contract's binary search algorithm to iterate correctly. Instead of being upheld locally, this trait is upheld by `reserveRangeInternal` within `OfferBase` as well as `reserveRange` in `BosonVoucher` indirectly.

The first enforcement is inferred by how new reserved ranges are always in incremental fashion within `OfferOrder` while the latter enforcement occurs by the `reserveRange` function of `BosonVoucher` disallowing a range to be reserved again for the same offer ID.

### Impact:

The binary search algorithm will cease functioning properly if the ascending order operational guarantee is not upheld which in turn would render premints inoperable. It is imperative that this type of guarantee is upheld locally due to the upgradeable nature of the Boson Protocol.

### Example:

```solidity
/**
 * @notice Reserves a range of vouchers to be associated with an offer
 *
 * Must happen prior to calling preMint
 * Caller must have PROTOCOL role.
 *
 * Reverts if:
 * - Start id is not greater than zero
 * - Offer id is already associated with a range
 *
 * @param _offerId - the id of the offer
 * @param _start - the first id of the token range
 * @param _length - the length of the range
 */
function reserveRange(
    uint256 _offerId,
    uint256 _start,
    uint256 _length
) external onlyRole(PROTOCOL) {
    // Make sure range start is valid
    require(_start > 0, INVALID_RANGE_START);

    // Get storage slot for the range
    Range storage range = _rangeByOfferId[_offerId];

    // Revert if the offer id is already associated with a range
    require(range.length == 0, OFFER_RANGE_ALREADY_RESERVED);

    // Store the reserved range
    range.offerId = _offerId;
    range.start = _start;
    range.length = _length;
    _rangeOfferIds.push(_offerId);

    emit RangeReserved(_offerId, range);
}
```

### Recommendation:

The security guarantee of an ascending order in the `_rangeOfferIds` should be validated locally rather than relying on proper sanitization by the caller as this type of validation delegation can lead to latent vulnerabilities. We advise the ascending order to be validated and enforced locally, ensuring the operational assumptions of the `BosonVoucher` are always upheld.

### LINKED ISSUE #504: [BVR-05M] Incorrect Invocation of Voucher Transfer Hook
## BVR-05M: Incorrect Invocation of Voucher Transfer Hook

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) |  | BosonVoucher.sol:L652-L653 |

### Description:

The `onVoucherTransferred` fallback in case the `_from` address is the `owner()` but the transfer of the NFT asset is not `committable` will invoke the `onVoucherTransferred` hook regardless of whether the `_to` address is zero or whether `_from` and `_to` are identical.

### Impact:

Currently, it is possible to bypass the security guarantees of `onVoucherTransferred` and invoke it with incorrect arguments as well as during burn operations if the `owner` is the `_from` address.

### Example:

```solidity
function _beforeTokenTransfer(
    address _from,
    address _to,
    uint256 _tokenId
) internal override {
    // Update the buyer associated with the voucher in the protocol
    // Only when transferring, not minting or burning
    if (_from == owner()) {
        if (_premintStatus.committable) {
            // Set the preminted token as committed
            _committed[_tokenId] = true;

            // If this is a transfer of preminted token, treat it differently
            address protocolDiamond = IClientExternalAddresses(BeaconClientLib._beacon()).getProtocolAddress();
            IBosonExchangeHandler(protocolDiamond).commitToPreMintedOffer(
                payable(_to),
                _premintStatus.offerId,
                _tokenId
            );
            delete _premintStatus;
        } else {
            // Already committed, treat as a normal transfer
            onVoucherTransferred(_tokenId, payable(_to));
        }
    } else if (_from != address(0) && _to != address(0) && _from != _to) {
        onVoucherTransferred(_tokenId, payable(_to));
    }
}
```

### Recommendation:

We advise the inner `if` case of the upper `if` clause validating the `owner` to be relocated to it (i.e. `if (_from == owner() && _premintStatus.committable)`), ensuring all conditions are correctly evaluated for the `onVoucherTransferred` hook execution. Additionally, given that `committable == true` infers that `silentMint` was invoked, the `_from` address is guaranteed to be the `owner()` in such a case. As such, the condition can be further optimized as only `_premintStatus.commitable == true`.

### LINKED ISSUE #508: [BVR-01C] Duplication of Data Point
## BVR-01C: Duplication of Data Point

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Gas Optimization](/appendix/finding-types#gas-optimization) |  | BosonVoucher.sol:L167 |

### Description:

The `_offerId` data point is used as a key in the `_rangeByOfferId` mapping and is additionally stored within the `Range` struct.

### Example:

```solidity
function reserveRange(
    uint256 _offerId,
    uint256 _start,
    uint256 _length
) external onlyRole(PROTOCOL) {
    // Make sure range start is valid
    require(_start > 0, INVALID_RANGE_START);

    // Get storage slot for the range
    Range storage range = _rangeByOfferId[_offerId];

    // Revert if the offer id is already associated with a range
    require(range.length == 0, OFFER_RANGE_ALREADY_RESERVED);

    // Store the reserved range
    range.offerId = _offerId;
    range.start = _start;
    range.length = _length;
    _rangeOfferIds.push(_offerId);

    emit RangeReserved(_offerId, range);
}
```

### Recommendation:

We advise it to either be omitted from the `Range` struct or to be filled in-memory whenever needed by callers as it is currently a redundant duplication of data as accessing the `Range` struct mandates knowledge of the `_offerId` it is associated with.

### LINKED ISSUE #509: [BVR-02C] Generic Typographic Mistake
## BVR-02C: Generic Typographic Mistake

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Code Style](/appendix/finding-types#code-style) |  | BosonVoucher.sol:L69 |

### Description:

The referenced line contains a typographical mistake or generic documentational error (i.e. copy-paste) that should be corrected.

### Example:

```solidity
// Tell is preminted voucher has already been _committed
```

### Recommendation:

We advise this to be done so to enhance the legibility of the codebase.

### LINKED ISSUE #510: [BVR-03C] Optimization of Binary Search Algorithm
## BVR-03C: Optimization of Binary Search Algorithm

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Gas Optimization](/appendix/finding-types#gas-optimization) |  | BosonVoucher.sol:L696-L709 |

### Description:

The custom binary search algorithm within `getPreMintStatus` evaluates whether each iteration of the search algorithm falls within the minted range of a particular reserved premint as well as within the reserved range of it.

### Example:

```solidity
function getPreMintStatus(uint256 _tokenId) internal view returns (bool committable, uint256 offerId) {
    // Not committable if _committed already or if token has an owner
    if (!_committed[_tokenId] && !_exists(_tokenId)) {
        // If are reserved ranges, search them
        uint256 length = _rangeOfferIds.length;
        if (length > 0) {
            // Binary search the ranges array
            uint256 low = 0; // Lower bound of search (array index)
            uint256 high = length; // Upper bound of search

            while (low < high) {
                // Calculate the current midpoint
                uint256 mid = (high + low) / 2;

                // Get the range stored at the midpoint
                Range storage range = _rangeByOfferId[_rangeOfferIds[mid]];

                // Get the beginning of the range once for reference
                uint256 start = range.start;

                if (start > _tokenId) {
                    // Split low and search again if target too high
                    high = mid;
                } else if (start + range.minted - 1 >= _tokenId) {
                    // Is token in target's minted range?

                    // It is committable if it has not been burned
                    if (_tokenId > range.lastBurnedTokenId) {
                        committable = true;
                        offerId = range.offerId;
                    }
                    break; // Found!
                } else if (start + range.length - 1 >= _tokenId) {
                    // No? Ok, is it in target's reserved range?
                    committable = false;
                    break; // Found!
                } else {
                    // No? It may be in a higher range
                    low = mid + 1;
                }
            }
        }
    }
}
```

### Recommendation:

Given that the minted range of a reserved premint is a subset case of its reserved range, we advise only the latter to be evaluated and its inner body to be expanded to validate whether the ID that falls within the reserved range is preminted, thus reducing the complexity and operational cost of the binary search algorithm in its worst case scenario as well as significantly optimizing the best case scenario as only one member of the `range` struct would be read on each binary search iteration instead of both `minted` and `length`.

### PR DESCRIPTION
**Pull request by zajck**
Fixes #500
Fixes #501 
Fixes #502
Fixes #503
Fixes #504 
Fixes #508 
Fixes #509 
FIxes #510

### CODE REVIEW

**Review by mischat — COMMENTED** [review]
This fixes : https://github.com/bosonprotocol/boson-protocol-contracts/issues/500

LGTM

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 1]
```diff
@@ -156,6 +157,7 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
     ) external onlyRole(PROTOCOL) {
         // Make sure range start is valid
         require(_start > 0, INVALID_RANGE_START);
+        require(_start <= type(uint256).max - _length, INVALID_RANGE_LENGTH);
```
Comment (mischat) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:164]: Is the reason we are using <= and not merely < because `type(uint256).max` is a reserved magic number ?

**Review by mischat — COMMENTED** [review]
Added a some question to this one, for my understanding.  This should fix https://github.com/bosonprotocol/boson-protocol-contracts/issues/501

**Review by mischat — COMMENTED** [review]
This fixes https://github.com/bosonprotocol/boson-protocol-contracts/issues/502 and LGTM

**Review by mischat — COMMENTED** [review]
this looks good to me. It fixes https://github.com/bosonprotocol/boson-protocol-contracts/issues/503

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 2]
```diff
@@ -651,25 +651,22 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
...
-
-                // If this is a transfer of preminted token, treat it differently
-                address protocolDiamond = IClientExternalAddresses(BeaconClientLib._beacon()).getProtocolAddress();
-                IBosonExchangeHandler(protocolDiamond).commitToPreMintedOffer(payable(_to), offerId, _tokenId);
-            } else {
-                // Already committed, treat as a normal transfer
-                onVoucherTransferred(_tokenId, payable(_to));
-            }
+        if (_premintStatus.committable) {
+            // If is commitable, invke commitToPreMintedOffer on the protocol
```
Comment (mischat) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:654]: I would move this up a line above the if statement, no ?

Also s/commitable/committable/

and 

s/invke/invoke/

**Review by mischat — COMMENTED** [review]
This fixes https://github.com/bosonprotocol/boson-protocol-contracts/issues/504

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 1 above)
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol:164]: `<=` allows that max unit is actually reached (even though practically will never be).
If we used `<`, the last range would need to end at `type(uint256).max-1`

So  `<=` actually makes `type(uint256).max` just completely ordinary number (while `<` would make it special).

**Review by mischat — COMMENTED** [review]
This looks good to me and addresses https://github.com/bosonprotocol/boson-protocol-contracts/issues/508

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 2 above)
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol:654]: I fixed the typos.

If I move it outside the if statement, then where to move first two comments of `else if`?
They both serve the same purpose, that's why I think it makes sense to have these comments where they are.

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 3]
```diff
@@ -651,25 +651,22 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
...
+            uint256 offerId = _premintStatus.offerId;
+            delete _premintStatus;
+
+            // Set the preminted token as committed
+            _committed[_tokenId] = true;
+
+            // If this is a transfer of preminted token, treat it differently
+            address protocolDiamond = IClientExternalAddresses(BeaconClientLib._beacon()).getProtocolAddress();
+            IBosonExchangeHandler(protocolDiamond).commitToPreMintedOffer(payable(_to), offerId, _tokenId);
         } else if (_from != address(0) && _to != address(0) && _from != _to) {
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol:666]: Yes, `_from == address(0)` is possible only during a mint.

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 4]
```diff
@@ -707,19 +707,19 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
...
+
+                        if (start + range.minted - 1 >= _tokenId) {
+                            // Has it been pre-minted?
 
-                        // It is committable if it has not been burned
-                        if (_tokenId > range.lastBurnedTokenId) {
-                            committable = true;
-                            offerId = currentOfferId;
+                            // It is committable if it has not been burned
+                            if (_tokenId > range.lastBurnedTokenId) {
```
Comment (mischat) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:717]: could we not merge the two if statements together ?

```if ((start + range.minted - 1 >= _tokenId) && (_tokenId > range.lastBurnedTokenId)) { ```

Does that make it more readable ?

**Review by mischat — COMMENTED** [review]
This should address https://github.com/bosonprotocol/boson-protocol-contracts/issues/510

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 4 above)
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol]: I merged them, makes sense.

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 2 above)
Comment (mischat) [reply contracts/protocol/clients/voucher/BosonVoucher.sol]: I guess consistency is the key, but yeah, i would have had them above, but it doesn't matter. 

If it is consistent and you are happy, then great.

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (mischat) [reply contracts/protocol/clients/voucher/BosonVoucher.sol:666]: thanks

**Review by mischat — APPROVED** [review]
This all looks good to me, thanks !

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 5]
```diff
@@ -154,8 +157,23 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
...
+        // Prevent reservation of an empty range
+        require(_length > 0, INVALID_RANGE_LENGTH);
+
+        // Prevent overflow in issueVoucher and preMint
+        require(_start <= type(uint256).max - _length, INVALID_RANGE_LENGTH);
+
+        // Make sure that ranges are in ascending order
+        if (_rangeOfferIds.length > 0) {
+            // Get latest registered range
+            Range storage lastRange = _rangeByOfferId[_rangeOfferIds[_rangeOfferIds.length - 1]];
```
Comment (anajuliabit) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:169]:
```suggestion
        uint256[] rangeOfferIds = _rangeOfferIds;
        if (rangeOfferIds.length > 0) {
            // Get latest registered range
            Range storage lastRange = _rangeByOfferId[rangeOfferIds[rangeOfferIds.length - 1]];
```
gas optimization

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 6]
```diff
@@ -633,26 +650,22 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
         address _to,
         uint256 _tokenId
     ) internal override {
-        // Update the buyer associated with the voucher in the protocol
-        // Only when transferring, not minting or burning
-        if (_from == owner()) {
```
Comment (anajuliabit) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:638]: Why did you remove this line? 
The only issue related to this that I found was https://github.com/bosonprotocol/boson-protocol-contracts/issues/504 and their suggestion looks different

**Review by anajuliabit — CHANGES_REQUESTED** [review]
Looks pretty good, just a few things

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 5 above)
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol]: I assume you meant "uint256[] **memory** rangeOfferIds = _rangeOfferIds;" This is not an optimization, since it copies all entries from storage to memory. Already when there are two elemens in the array, that becomes more expensive than current implementation. Whit multiple ranges it gets worse.

Instead I only stored the length to variable and thus save a bit of gas.

File: contracts/protocol/clients/voucher/BosonVoucher.sol (hunk 6 above)
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol:638]: Implementation is exactly what they suggest.

First they suggest changing `if (_from == owner())` to  `if (_from == owner() && _premintStatus.committable)`

So this is true if both are true, but as soon as `_premintStatus.committable == true`, it always means `_from == owner()`. So it completely sufficient to test only `_premintStatus.committable == true` as they suggest at very end. I just didn't do explicit check of equality `if(_premintStatus.committable == true)` since commitable is already bool.

**Review by anajuliabit — DISMISSED** [review]

**Review by anajuliabit — APPROVED** [review]

**Review by mischat — APPROVED** [review]
The only change since my last review was a typo. 

Thanks, this looks good to me
